### DISCUSSION TITLE
Node text locales

### DISCUSSION THREAD

**Discussion by GrimPixel** [IndiePython/nodezator#54] [Feature Request]
There are texts on the nodes. There can be localisations on them. 
A node pack can provide locales for node texts and the program can provide the ability to choose.

**Comment by KennedyRichard** [comment]
I agree. Much has to be discussed regarding how this will work, but the core idea seems great. I accept it :heavy_check_mark: 

Again, I cannot tackle this at the moment, but once the most urgent stuff is implement, I do want to implement something like this. I also want to research what other Python projects do regarding internationalization of their documentation.

I know that a tool called Sphix is used a lot for documentation generation in Python projects and that it supports internationalization/localization but I'd like to implement a simpler and more beginner-friendly version for Nodezator. Something just like you described, by which users can simply specify additional text in other languages and people looking into the node's source have the option to see the additional content in another language(s) provided by the node's creator.

**Reply by GrimPixel** [reply]
Glad to know that you are going to do it.
In my own Python program “Numeration”, YAML files are used to store locales. In the settings file, the locale can be specified by its code. The program gets the locale code, reads the keys and returns their values.

**Comment by KennedyRichard** [comment]
Already listed the feature request in the list of planned/requested features #13

**Comment by GrimPixel** [comment]
In my own repository, English node names can be too long. I am thinking about using Chinese instead. Then English will be available through locales.
Nah. This is not the direction I am going.

**Comment by GrimPixel** [comment]
I am applying the approach mentioned at https://github.com/IndiePython/nodezator/discussions/73, which functions almost the same as how “locales” do, but it's using a list instead of a dictionary. It is a “globalisation” of formulae. This is available at https://codeberg.org/GrimPixel/Nodezator-Equationary

I think the steps of use could be like this:
1. search node names and node's “labels” through the search bar mentioned at https://github.com/IndiePython/nodezator/discussions/27.
2. apply the labels in that set if that result is selected.

To develop this, I think the steps are:
1. read node names and node's “labels”
2. replace the original names of input/output parameters with the labels, if the result selected is a label set (output first, input last)
3. display the names for the input/output

**Comment by KennedyRichard** [comment]
The names of the node pack, category folders and node script folders must all be valid Python identifiers (in other words, if you call`str.isidentifier` on their name in the interpreter it must return True), otherwise the name can't be used. This is not a limitation I impose, this is imposed by the Python language and interpreter themselves, I have no control over this, so you must change those names to names accepted by the interpreter (valid Python identifiers).

For instance, the name of the node pack cannot have the `-` character. So, instead of `Nodezator-Equationary` you should use `NodezatorEquationary`, `Nodezator_Equationary` or `Equationary`, etc (though I recommend using lowercase letters, like most Python packages do). The name of the category folders, 1, 2, 3 ,4, that you used  aren't allowed either because Python identifiers must start with a letter or `_` character, not with a number. The names of the node script folder are also using forbidden characters.

If you try executing your node scripts by themselves in Python the interpreter should point out the errors so you can easily fix them.

**Reply by GrimPixel** [reply]
Thank you for pointing out the hyphen problem. It is an underscore now.

It seems that Python treats fullwidth characters as halfwidth. It is not allowed to use the invalid characters even if they are fullwidth.
